File: Ubapp/RealTime/EnemyEntity.swift
import SpriteKit
import GameplayKit

final class EnemyEntity: GKEntity {
    let sprite: SpriteComponent
    let agent: AgentComponent
    weak var target: SKNode?
    var chaseProxy: GKAgent2D?
    private(set) var stateMachine: GKStateMachine!

    init(target: SKNode) {
        sprite = SpriteComponent(color: .systemGray, radius: 14)
        agent = AgentComponent()
        self.target = target
        super.init()
        agent.sprite = sprite.node
        agent.maxSpeed = 110
        agent.maxAcceleration = 200
        addComponent(sprite)
        addComponent(agent)
        stateMachine = GKStateMachine(states: [WanderState(enemy: self), ChaseState(enemy: self)])
        stateMachine.enter(WanderState.self)
    }

    required init?(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }

    override func update(deltaTime seconds: TimeInterval) {
        super.update(deltaTime: seconds)
        stateMachine.update(deltaTime: seconds)
    }
}

File: android/app/src/main/java/com/example/ubapp/realtime/EnemyEntity.kt
package com.example.ubapp.realtime

import android.graphics.Color

class EnemyEntity(
    x: Float,
    y: Float,
    val target: SpriteComponent,
    private val worldWidth: () -> Float,
    private val worldHeight: () -> Float
) {
    val entity = Entity()
    val sprite = SpriteComponent(x, y, radius = 16f, color = Color.GRAY)
    val agent = AgentComponent(maxSpeed = 140f, maxAccel = 280f)

    val stateMachine: StateMachine

    init {
        entity.add(sprite).add(agent)
        val wander = WanderState(this, worldWidth, worldHeight)
        val chase = ChaseState(this)
        stateMachine = StateMachine(wander)
        wander.peer = chase
        chase.peer = wander
    }

    fun update(dt: Float) {
        stateMachine.update(dt)
        entity.update(dt)
    }
}
